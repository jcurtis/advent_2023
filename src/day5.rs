use itertools::Itertools;
use lazy_static::lazy_static;
use rayon::prelude::*;
use regex::Regex;
use std::collections::HashMap;

type Seed = u64;
type Mapping = String;

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
struct Map {
    to: String,
    destination_start: u64,
    source_start: u64,
    range_length: u64,
}

type Maps = HashMap<Mapping, Vec<Map>>;

#[derive(Debug)]
pub struct Almanac {
    seeds: Vec<Seed>,
    maps: Maps,
}

lazy_static! {
    static ref MAPPING_RE: Regex = Regex::new(r"(\w+)-to-(\w+) map:").unwrap();
}

pub fn generator(input: &str) -> Almanac {
    let seeds = input.lines().next().unwrap();
    let seeds = seeds
        .split(' ')
        .skip(1)
        .map(|num| num.parse::<Seed>().unwrap())
        .collect_vec();

    let maps: Maps = input
        .split("\n\n")
        .skip(1)
        .map(|section| {
            let mapping = MAPPING_RE.captures(section).unwrap();
            (
                mapping[1].to_string(),
                section
                    .lines()
                    .skip(1)
                    .map(|line| {
                        let nums = line
                            .split(' ')
                            .map(|num| num.parse::<u64>().unwrap())
                            .collect_vec();
                        Map {
                            to: mapping[2].to_string(),
                            destination_start: nums[0],
                            source_start: nums[1],
                            range_length: nums[2],
                        }
                    })
                    .collect_vec(),
            )
        })
        .collect();

    Almanac { seeds, maps }
}

fn seed_location(seed: Seed, almanac: &Almanac) -> u64 {
    let location = "location".to_string();
    let mut cur_map = "seed".to_string();
    let mut cur_index = seed;
    let mut count = 0;

    // println!("0 - soil {cur_index}");
    while cur_map.ne(&location) && count < 100 {
        let found_maps = almanac.maps.get(&cur_map).unwrap();
        let found_map = found_maps.iter().find(|map| {
            cur_index >= map.source_start && cur_index < map.source_start + map.range_length
        });
        cur_index = if let Some(map) = found_map {
            let offset = cur_index - map.source_start;
            map.destination_start + offset
        } else {
            cur_index
        };

        cur_map = found_maps[0].to.clone();
        count += 1;

        // println!("{count} - {} {cur_index}", &cur_map);
    }

    cur_index
}

pub fn part_1(almanac: &Almanac) -> Seed {
    almanac
        .seeds
        .iter()
        .map(|seed| seed_location(*seed, almanac))
        .min()
        .unwrap()
}

pub fn part_2(almanac: &Almanac) -> u64 {
    almanac
        .seeds
        .chunks(2)
        .map(|range| {
            let range = range[0]..(range[0] + range[1]);
            println!("checking range {:?}", &range);
            range
                .into_par_iter()
                .map(|seed| seed_location(seed, almanac))
                .min()
                .unwrap()
        })
        .min()
        .unwrap()
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    lazy_static! {
        static ref ALMANAC: Almanac = generator(&INPUT);
    }

    #[test]
    fn test_generator() {
        assert_eq!(ALMANAC.seeds, vec![79, 14, 55, 13]);
        let sorted_map = ALMANAC.maps.iter().sorted();
        assert_debug_snapshot!(&sorted_map);
    }

    #[test]
    fn test_seed_location() {
        assert_eq!(seed_location(79, &ALMANAC), 82);
        assert_eq!(seed_location(14, &ALMANAC), 43);
        assert_eq!(seed_location(55, &ALMANAC), 86);
        assert_eq!(seed_location(13, &ALMANAC), 35);
    }

    #[test]
    fn test_part_1() {
        assert_eq!(part_1(&ALMANAC), 35);
    }

    #[test]
    fn test_part_2() {
        assert_eq!(part_2(&ALMANAC), 46);
    }

    static INPUT: &str = "seeds: 79 14 55 13

seed-to-soil map:
50 98 2
52 50 48

soil-to-fertilizer map:
0 15 37
37 52 2
39 0 15

fertilizer-to-water map:
49 53 8
0 11 42
42 0 7
57 7 4

water-to-light map:
88 18 7
18 25 70

light-to-temperature map:
45 77 23
81 45 19
68 64 13

temperature-to-humidity map:
0 69 1
1 0 69

humidity-to-location map:
60 56 37
56 93 4
";
}
