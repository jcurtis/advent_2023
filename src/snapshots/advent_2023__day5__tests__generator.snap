---
source: src/day5.rs
expression: "&sorted_map"
---
IntoIter(
    [
        (
            "fertilizer",
            [
                Map {
                    to: "water",
                    destination_start: 49,
                    source_start: 53,
                    range_length: 8,
                },
                Map {
                    to: "water",
                    destination_start: 0,
                    source_start: 11,
                    range_length: 42,
                },
                Map {
                    to: "water",
                    destination_start: 42,
                    source_start: 0,
                    range_length: 7,
                },
                Map {
                    to: "water",
                    destination_start: 57,
                    source_start: 7,
                    range_length: 4,
                },
            ],
        ),
        (
            "humidity",
            [
                Map {
                    to: "location",
                    destination_start: 60,
                    source_start: 56,
                    range_length: 37,
                },
                Map {
                    to: "location",
                    destination_start: 56,
                    source_start: 93,
                    range_length: 4,
                },
            ],
        ),
        (
            "light",
            [
                Map {
                    to: "temperature",
                    destination_start: 45,
                    source_start: 77,
                    range_length: 23,
                },
                Map {
                    to: "temperature",
                    destination_start: 81,
                    source_start: 45,
                    range_length: 19,
                },
                Map {
                    to: "temperature",
                    destination_start: 68,
                    source_start: 64,
                    range_length: 13,
                },
            ],
        ),
        (
            "seed",
            [
                Map {
                    to: "soil",
                    destination_start: 50,
                    source_start: 98,
                    range_length: 2,
                },
                Map {
                    to: "soil",
                    destination_start: 52,
                    source_start: 50,
                    range_length: 48,
                },
            ],
        ),
        (
            "soil",
            [
                Map {
                    to: "fertilizer",
                    destination_start: 0,
                    source_start: 15,
                    range_length: 37,
                },
                Map {
                    to: "fertilizer",
                    destination_start: 37,
                    source_start: 52,
                    range_length: 2,
                },
                Map {
                    to: "fertilizer",
                    destination_start: 39,
                    source_start: 0,
                    range_length: 15,
                },
            ],
        ),
        (
            "temperature",
            [
                Map {
                    to: "humidity",
                    destination_start: 0,
                    source_start: 69,
                    range_length: 1,
                },
                Map {
                    to: "humidity",
                    destination_start: 1,
                    source_start: 0,
                    range_length: 69,
                },
            ],
        ),
        (
            "water",
            [
                Map {
                    to: "light",
                    destination_start: 88,
                    source_start: 18,
                    range_length: 7,
                },
                Map {
                    to: "light",
                    destination_start: 18,
                    source_start: 25,
                    range_length: 70,
                },
            ],
        ),
    ],
)
